src/script/media/VideoBackgroundBlur.ts

```diff
@@ -62,111 +57,68 @@ function startBlurProcess(
-      originalPixels[baseIndex + 3] = blurredPixels[baseIndex + 3];
-    }
-  }
-}
-
-async function createSegmenter(): Promise<ImageSegmenter> {
+async function createSegmenter(canvas: HTMLCanvasElement): Promise<ImageSegmenter> {
   const video = await FilesetResolver.forVisionTasks('./mediapipe/wasm');
   return ImageSegmenter.createFromOptions(video, {
     baseOptions: {
       modelAssetPath: QualitySettings.segmentationModel,
       delegate: 'GPU',
     },
+    canvas,
     runningMode: 'VIDEO',
     outputCategoryMask: false,
     outputConfidenceMasks: true,
   });
 }
 
-export async function applyBlur(originalStream: MediaStream): Promise<{stream: MediaStream; release: () => void}> {
+/**
+ * Will create a new MediaStream that will both segment each frame and apply a blur effect to the background.
+ * @param originalStream the stream that contains the video that needs background blur
+ * @returns a promise that resolves to an object containing the new MediaStream and a release function to stop the blur process
+ */
+export async function applyBlur(stream: MediaStream): Promise<{stream: MediaStream; release: () => void}> {
   // Create a video element to display the webcam feed
   const videoEl = document.createElement('video');
-  // Create a canvas element to apply the blur effect
-  const canvasEl = document.createElement('canvas');
-  // Get the 2D context of the canvas
-  const ctx = canvasEl.getContext('2d', {willReadFrequently: true})!;
+  // Create a canvas element that will be to draw the blurred frames
   // Store the video dimensions
   const videoDimensions = {width: 0, height: 0};
 
-  const segmenter = await createSegmenter();
-
-  videoEl.srcObject = originalStream.clone();
+  const originalStream = stream.clone();
+  videoEl.srcObject = originalStream;
   videoEl.onloadedmetadata = () => {
     // Ensure metadata is loaded to get video dimensions
     videoDimensions.width = videoEl.videoWidth || 1240;
     videoDimensions.height = videoEl.videoHeight || 720;
-    canvasEl.width = videoDimensions.width;
-    canvasEl.height = videoDimensions.height;
-
     videoEl.play().catch(error => console.error('Error playing the video: ', error));
   };
 
   return new Promise(resolve => {
-    videoEl.onplay = () => {
-      const stopBlurProcess = startBlurProcess(segmenter, ctx, videoEl, videoDimensions);
-      const videoStream = canvasEl.captureStream(QualitySettings.framerate).getVideoTracks()[0];
+    videoEl.onplay = async () => {
+      const glContext = document.createElement('canvas');
+      glContext.height = videoDimensions.height;
+      glContext.width = videoDimensions.width;
+
+      const gl = initShaderProgram(glContext, videoDimensions);
+      const segmenter = await createSegmenter(glContext);
+
+      const stopBlurProcess = startBlurProcess(segmenter, gl, videoEl, videoDimensions);
+      const videoStream = glContext.captureStream(QualitySettings.framerate).getVideoTracks()[0];
```

Comment:
**suggestion**: How about an extra check if the videoStream is defined? Maybe we should add `--noUncheckedIndexedAccess` flag in our tsconfig in general?

Reply:
humm that's good suggestion. We shall try to add it in a follow up for sure (it might need some code adjustment)